### PR TITLE
Possible typeclass

### PR DESCRIPTION
**Pull request by cquiroz**
This is really just an experiment with the new encoding.
Please do thorough check as I'm not totally familiar with this style

### CODE REVIEW

File: kernel/shared/src/main/scala/monocle/Optional.scala [hunk 1]
```diff
@@ -28,4 +29,7 @@ object Optional {
 
   def tailOption[S](implicit ev: Cons[S]): Optional[S, S] =
     ev.tailOption
+
+  def possible[S, A](implicit ev: Possible.Aux[S, A]): Optional[S, A] =
```
Comment (cquiroz) [inline kernel/shared/src/main/scala/monocle/Optional.scala]: Is this really useful?
A related question why is `headOption[S, A]: Optional[S, A]` while `tailOption[S]: Optional[S, S]`?

File: kernel/shared/src/main/scala/monocle/function/Possible.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+object Possible {
+  type Aux[S, A0] = Possible[S] { type A = A0 }
+
+  def apply[S, A0](optional: Optional[S, A0]): Aux[S, A0] = new Possible[S] {
+    type A = A0
+    override val possible: Optional[S, A0] = optional
+  }
+
+  implicit def optionPossible[A0](implicit ev: Optional[Option[A0], A0]): Possible[Option[A0]] =
```
Comment (cquiroz) [inline kernel/shared/src/main/scala/monocle/function/Possible.scala]: I guess we don't want this `implicit` do we?

File: kernel/shared/src/main/scala/monocle/Optional.scala (hunk 1 above)
Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/Optional.scala]: I think it is. It is a way to summon a `possible` instance such as we can do:
```scala
foo compose Optional.possible
```

I believe we can offer a much better syntax in another module. I will send a proof of concept PR soon

File: kernel/shared/src/main/scala/monocle/function/Possible.scala (hunk 2 above)
Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/function/Possible.scala]: yeah here we should define an instance of the typeclass Possible for `Option[A]`. So the signature is most likely going to be:
```scala
implicit def option[A]: Possible.Aux[Option[A], A] = apply(...) 
```

File: kernel/shared/src/main/scala/monocle/Optional.scala (hunk 1 above)
Comment (cquiroz) [reply kernel/shared/src/main/scala/monocle/Optional.scala]: Do you think I should remove it?

Comment (julien-truffaut) [reply kernel/shared/src/main/scala/monocle/Optional.scala]: ah no, I think it is useful

**Comment by cquiroz** [comment]
Oh I did something bad with a rebase I'll redo the PR

**Pull request closed without merging**